pint_server/app.py

```diff
@@ -452,11 +688,27 @@ def assert_valid_provider(provider):
         abort(Response('', status=404))
 
 
+def assert_valid_provider_region(provider, region):
+    provider_regions = query_provider_regions(provider)
```

Comment:
This will result in 2 queries instead of one. Not sure what the performance impact is for high volume calls. I think we don't need this as the query will result in 404 anyway for an invalid region.

Reply:
The first query, for an empty cache, will trigger the DB operations to determine the provider specific regions list, whereas the subsequent runs will get the value from the cache.

The use of the hash argument in the @lru_cache'd function does incur additional DB activity... But it is needed to ensure that we don't incorrectly re-use stale regions data...

The Python version we are currently based on doesn't really have a decent TTL caching mechanism; newer Python releases have something more useful in that regard.

An alternative that doesn't lead to DB operations (which is what I started out with) would be to have the get_provider_hash() method just return a timestamp value rounded to the end of the nearest X second boundary window, ensuring that cache entries will expire once we move to a new boundary window...

However that approach runs the risk of inappropriately using stale cached data if for example we use a 1 minute boundary, and a query comes in at 1 second into the window, and then a DB update modifies the region details for that provider soon afterwards, say 5 seconds into the minute... Until the end of the minute boundary window, e.g. another 55 seconds, the cached data will be used which is stale...

So the hashing of the data version details is a compromise to ensure that we don't use stale data, while providing the means to cache data that might be re-used which could be a bit expensive to generate in the first place...

I'm happy to switch back the timestamp boundary window idea if this seems too expensive, but we probably will need to keep the boundary window pretty small to avoid prolonged re-use of stale cached data, which negates most of the benefit of the caching.